ASDPLNG-134: Standard Profile for Web Server

billglick opened the pull request:
```
add in apache::mods, letsencrypt, etc.
Update default firewall subnets, along with examples
Add in profile_website::modules::php::ini_file parameter
Move apache parameters including modules into control repo
Specifically include various modules
include apache::vhost
Update letsencrypt to use all domains from default vhost
Setup rhel8 to use php:7.4 dnf module
Ensure kerberos http principal & keytab
Update README will full working hiera example
```

See https://jira.ncsa.illinois.edu/browse/ASDPLNG-134

The goal of this module it to standup a standard HTTPd web server with all of our NCSA best practices applied. Out of the box it also installs:
- LetsEncrypt SSL certificate that is renewed via cron
- PHP 7.4 (on RHEL8)
- mod_auth_gssapi & mod_ldap to support kerberos authn with ldap authz
- httpd logging to syslog and thus log forwarding

This is currently being tested on `asd-test-web1`.

billglick committed:
ASDPLNG-134: Standard Profile for Web Server

Add in apache::mods, letsencrypt, etc.
Update default firewall subnets, along with examples
Add in profile_website::modules::php::ini_file paramter
Move apache paraameters including modules into control repo
Specifically include various modules
Include apache::vhost
Update letsencrypt to use all domains from default vhost
Setup rhel8 to use php:7.4 dnf module
Ensure kerberos http principal & keytab
Update README will full working hiera example
Add monitoring class for generating exported resources to profile_monitoring
Add custom facts to list installed ssl_certificates & ssl_keys
Use temporary vhost if SSL cert files do not exist yet to allow LetsEncrypt to work

billglick commented:
I have now added in the Puppet exported resources to trigger telegraf monitoring of the sslcert & website. I had forgotten about that until this morning, but it is now added, tested, verified, and rebased.

This is ready for review now.

billglick commented:
I setup a fresh `asd-test-web2` host and have discovered an issue where LetsEncrypt needs a temporary web server up and running in order to generate the initial SSL certificate. But currently the default website references use of the LetsEncrypt SSL certificates in order to start up.

We need to figure out a way to start up a temporary website if we do not already have SSL certificates generated, then switch to using the LetsEncrypt certificates once LetsEncrypt has created them. I do not have an idea yet to resolve this.

But perhaps a tentative review would still be useful before we address this issue. That review might even lead to an idea to address this issue.

billglick commented:
Although it isn't pretty, I have resolved the issue with needing a temporary, working `apache::vhost` to allow `letsencrypt` to create the initial certificates. It now takes 2 initial puppet runs to get fully setup:
1. The first run uses a temporary non-ssl apache vhost to allow Lets Encrypt to validate the website during the certificate creation.
2. On the 2nd run if we find the expected SSL certificate is installed, then it switches to using the apache vhost from the lookup date.

@andylytical This fix includes custom facts that list the SSL certificates & keys installed on the host. The custom facts have a line that `pdk-validate` is complaining is too long. I need help update the 2 facter scripts to pass pdk validation.

andylytical reviewed (changes requested):
## KERBEROS.PP
- Aren't these all in standard, pre-defined places (guaranteed by profile_kerberos)?
  - `$host_keytab_file`
  - `$kadmin_file`
- Can `$realm` be looked up via:
    - Hiera "lookup" from profile_kerberos?
    - Custom fact?
      - See also: https://github.com/ncsa/puppet-sshd/blob/main/lib/augeasfacter/sssd_info.conf

- Typo on line 43 `needst`

- What is "http_keytab_file"?
  - Why not hardcode this location?
  - If it's managed by Puppet, will it ever need to go somewhere else?

- What is the purpose of `exec { 'http_keytab_is_missing':` ?
  - Doesn't `exec { 'ensure_http_keytab'` create the keytab (or fail) ?

## MONITORING.PP
- Would like to chat about exported resources (I don't get them so I'm not a useful reviewer here.)

## PHP.PP
- Can `ensure_dnf_php_module_7.4` be done with puppet:yum ?
- Should php version be a parameter?

## SSL.pp
- Replace `$::fqdn` with `$facts['fqdn']` (explicit indication this is a reference to a host fact)

## VHOST.pp
- Fact references should use the `$facts[]` hash.
- How to ensure that temporary vhost get's removed on a second / subsequent run?
- What if puppet never runs a second time (for some reason), could the temporary vhost be a security issue?

## LIB/FACTER/*.RB
- Replace the BASH code with Ruby. See also:
  - https://stackoverflow.com/questions/1755665/get-names-of-all-files-from-a-folder-with-ruby
  - https://brandonrubin.me/2013/07/29/using-ruby-find-to-look-for-files/
  - https://ruby-doc.org/stdlib-2.7.1/libdoc/find/rdoc/Find.html

billglick commented:
Some comments on the questions posed:

> ## KERBEROS.PP
> * Aren't these all in standard, pre-defined places (guaranteed by profile_kerberos)?
>   
>   * `$host_keytab_file`
>   * `$kadmin_file`

Our `control_repo` template does not use `profile_kerberos`. We would need to update `profile_system_auth` to parameterize these values. I think that is a great idea, but I think that should be an enhancement issue of this profile module.

> * Can `$realm` be looked up via:
>   
>   * Hiera "lookup" from profile_kerberos?
>   * Custom fact?

Probably via a custom fact. But again, that fact should live in a different profile, `profile_system_auth`.
     
>     * See also: https://github.com/ncsa/puppet-sshd/blob/main/lib/augeasfacter/sssd_info.conf

I do not understand why this file above is referenced here.

> * What is "http_keytab_file"?

This file could/should be referenced by an `apache::vhost` hash that would be set outside of this profile. I was trying to keep it flexible so that it could be changed and/or looked up.

>   * Why not hardcode this location?
>   * If it's managed by Puppet, will it ever need to go somewhere else?

Yes, in an `apache::vhost` from hiera, likely in the control repo or another profile.

> * What is the purpose of `exec { 'http_keytab_is_missing':` ?
>   
>   * Doesn't `exec { 'ensure_http_keytab'` create the keytab (or fail) ?

I don't want to run the kerberos query in the `ensure_http_principal` `unless` command on every Puppet run. So, by adding `exec { 'http_keytab_is_missing':`, we are only triggering the other 2 execs if the keytab is missing. All other variations that I had tried of just having the last 2 `execs` resulted in the `ensure_http_principal` `unless` command running on every puppet run. Would an array of `unless` commands work so there so that the first checked if the keytab file was missing, and then only run the 2nd command of querying kerberos if the file was missing?
 
> ## PHP.PP
> * Can `ensure_dnf_php_module_7.4` be done with puppet:yum ?
> * Should php version be a parameter?

Yes, but it will only work for RHEL 8 here. We can't guarantee the version of PHP for EL 7, nor for CentOS 8, using this approach. And only certain version numbers.
I suggest a feature enhancement of incorporating using a one of the PHP puppet modules to handle this.

> ## LIB/FACTER/*.RB
> * Replace the BASH code with Ruby. See also:
>   
>   * https://stackoverflow.com/questions/1755665/get-names-of-all-files-from-a-folder-with-ruby
>   * https://brandonrubin.me/2013/07/29/using-ruby-find-to-look-for-files/
>   * https://ruby-doc.org/stdlib-2.7.1/libdoc/find/rdoc/Find.html

Is this a requirement? These finds are rather complicated.

billglick commented:
@andylytical  I think I've addressed all of your issues, except those that I will note below.

I have not yet rebased, so that it should be a bit easier to see the commits that address the issues you brought up.


> ## KERBEROS.PP

I removed all of the optional settings and am just letting the kadmin command use all the defaults. That lets it pickup the default realm, keytab, etc. and is much cleaner in this module.

And improved the execs so that there are only 2 and they are a bit smarter. I changed my mind about letting kadmin query kerberos on every puppet run. There shouldn't be any harm in that, and now it will recreate the HTTP keytab if the HTTP principal gets recreated.


> ## MONITORING.PP
> * Would like to chat about exported resources (I don't get them so I'm not a useful reviewer here.)

I'd be glad to chat more about this, but seeing actual examples helped me understand it. 

See https://puppet.com/docs/puppet/7/lang_exported.html & https://puppet.com/docs/puppet/7/lang_collectors.html

Exported resources just define puppet resource statements that are **`exported`** to PuppetDB, and then can be **`collected`** and applied by other hosts and/or classes. See the following examples of how we are doing this with exporting `ping_check` resources from each host, then collecting and applying those resources on the monitoring server:
- https://github.com/ncsa/puppet-profile_monitoring/blob/main/manifests/register_ping_check.pp
- https://github.com/ncsa/puppet-profile_monitoring/blob/main/manifests/telegraf_ping_check.pp


> ## PHP.PP
> * Can `ensure_dnf_php_module_7.4` be done with puppet:yum ?

I think `module` is a command specific to DNF. See https://dnf.readthedocs.io/en/latest/command_ref.html#module-command
I do not see a way to do manage DNF modules with `puppet/yum`.

> * Should php version be a parameter?

Lets defer this as new issue for an enhancement.


> ## VHOST.pp
> * How to ensure that temporary vhost get's removed on a second / subsequent run?

`puppet/apache`'s default setting it to purge any vhost configurations that it does not manage, so it gets removed when it uses the looked up vhosts.

> * What if puppet never runs a second time (for some reason), could the temporary vhost be a security issue?

I do not see a security issue. The temporary vhost is still applying NCSA Security's best practices, but it just is not a HTTPS site.

andylytical commented on manifests/kerberos.pp:
```diff
@@ -0,0 +1,42 @@
+
+  # KRB COMMAND STRINGS
+  $kadmin_query = "kadmin -k -p host/${::fqdn} -q"
+  $http_principal = "HTTP/${::fqdn}"
```
Replace `$::fqdn` with `$facts['fqdn']`

andylytical commented on README.md:
````diff
@@ -5,42 +5,111 @@
 ```
-include ::profile_website
+apache::default_ssl_cert: "/etc/letsencrypt/live/%{::fqdn}/fullchain.pem"
+apache::default_ssl_key: "/etc/letsencrypt/live/%{::fqdn}/privkey.pem"
````
Replace `%{::fqdn}` with `%{facts.fqdn}`

andylytical commented on README.md:
```diff
@@ -5,42 +5,111 @@
+apache::default_vhost: false
+apache::default_ssl_vhost: false
+apache::default_ssl_cert: "/etc/letsencrypt/live/%{::fqdn}/fullchain.pem"
+apache::default_ssl_key: "/etc/letsencrypt/live/%{::fqdn}/privkey.pem"
```
Replace `%{::fqdn}` with `%{facts.fqdn}`

andylytical commented on README.md:
```diff
@@ -5,42 +5,111 @@
+    docroot: "/var/www/html"
+    error_log_pipe: "|/bin/sh -c
+      '/usr/bin/tee
+      -a /var/log/httpd/%{::fqdn}-nossl_error.log'
```
Replace `%{::fqdn}` with `%{facts.fqdn}`

andylytical commented on README.md:
```diff
@@ -5,42 +5,111 @@
+    docroot: "/var/www/html"
+    error_log_pipe: "|/bin/sh -c
+      '/usr/bin/tee
+      -a /var/log/httpd/%{::fqdn}-ssl_error_ssl.log'
```
Replace `%{::fqdn}` with `%{facts.fqdn}`

andylytical commented on manifests/monitoring.pp:
```diff
@@ -0,0 +1,40 @@
+    ensure   => 'present',
+    after    => 'urls',
+    line     => "    \"https://${::fqdn}\",",
+    match    => $::fqdn,
```
Replace `$::fqdn` with `$facts['fqdn']`

andylytical commented on manifests/php.pp:
```diff
@@ -0,0 +1,59 @@
+  Hash[String, String] $ini_settings,
+) {
+
+  if ($::osfamily == 'RedHat' and $::operatingsystemmajrelease >= '8') {
```
Replace `$::osfamily` -> `$facts['os']['family']`

Replace `$::operatingsystemmajrelease` -> `$facts['os']['release']['major']`

andylytical commented on manifests/vhost.pp:
```diff
@@ -0,0 +1,65 @@
+  }
+  elsif (
+    $default_ssl_cert in $::ssl_certificates
+    and $default_ssl_key in $::ssl_keys
```
Replace `$::ssl_certificates` with `$facts['ssl_certificates']`

Replace `$::ssl_keys` with `$facts['ssl_keys']`

andylytical reviewed (changes requested):
Updates look really good. I especially like `kerberos.pp` and the two custom fact files; they are very clean and simple now!

A few more suggestions below, all related to fact variable references using the `$facts[]` hash.

I see no need for another review if these are the only changes made.
